rewriter/GenCallAsm.cpp

```diff
@@ -280,21 +280,309 @@ static std::string sig_string(const CAbiSignature &sig,
   return ss.str();
 }
 
+static void call_string(
+    const std::optional<std::string> target_name,
+    WrapperKind kind,
+    AsmWriter &aw,
+    Arch arch) {
+
+  add_comment_line(aw, "Call wrapped function");
+
+  if (arch == Arch::X86) {
+    if (!target_name) {
+      // indirect call
+      assert(kind == WrapperKind::IndirectCallsite);
+      add_asm_line(aw, "call *%r12");
+    } else {
+      // direct call
+      add_asm_line(aw, "call "s + target_name.value());
+    }
+  } else if (arch == Arch::Aarch64) {
+    if (!target_name) {
+      // indirect call
+      assert(kind == WrapperKind::IndirectCallsite);
+      add_asm_line(aw, "br "s + target_name.value());
+    } else {
+      // direct call
+      add_asm_line(aw, "b "s + target_name.value());
+    }
+  }
+}
+
+static AsmWriter get_asmwriter(bool as_macro) {
+  std::string terminator = {};
+  // Code generated as a macro needs to terminate each line with '\'
+  if (as_macro) {
+    terminator = "\\"s;
+  }
+  return {.ss = {}, .terminator = terminator};
+}
+
+static void switch_stacks(AsmWriter &aw, uint32_t caller_pkey, uint32_t target_pkey, Arch arch) {
+  if (arch == Arch::X86) {
+    // Save the old frame pointer and set the frame pointer for the call gate
+    add_asm_line(aw, "pushq %rbp");
```

Comment:
This is part of the function prologue, not exactly related to stack switching so it might make sense to move it out of this function. Especially if that means you can reuse this function for the second time we generate the stack switching code.

Reply:
Moved these into their own function `emit_save_regs`.